pkg/socketlb/cgroup.go

```diff
@@ -0,0 +1,240 @@
+	}
+
+	// Successfully created and pinned bpf_link.
+	return nil
+}
+
+// exclusiveCgroupAttach attaches prog after first detaching all preexisting
+// programs from prog's corresponding cgroup hook.
+//
+// The detach operation can fail if one of the preexisting programs is attached
+// through a bpf_link not managed by Cilium.
+func exclusiveCgroupAttach(prog *ebpf.Program, name, cgroupRoot string) (link.Link, error) {
+	if err := searchAndDetach(name, cgroupRoot, nil); err != nil {
+		return nil, fmt.Errorf("detach program: %w", err)
+	}
+
+	l, err := link.AttachCgroup(link.CgroupOptions{
+		Path:    cgroupRoot,
+		Attach:  attachTypes[name],
+		Program: prog,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("attach program %s: %w", name, err)
+	}
+
+	return l, nil
+}
+
+// detachCgroup detaches a program with the given name from cgroupRoot. Attempts
+// to open a pinned link with the given name from directory pinPath first,
+// falling back to PROG_DETACH if no pin is present.
+func detachCgroup(name, cgroupRoot, pinPath string) error {
+	pin := filepath.Join(pinPath, name)
+	l, err := link.LoadPinnedLink(pin, &ebpf.LoadPinOptions{})
+	if err == nil {
+		if err := l.Unpin(); err != nil {
+			return fmt.Errorf("unpin link %s: %w", pin, err)
+		}
+		if err := l.Close(); err != nil {
+			return fmt.Errorf("close link %s: %w", name, err)
+		}
+		return nil
+	}
+
+	// No bpf_link pin found, search and detach prog_attach progs.
+	log.Debugf("no pinned link '%s', querying cgroup", pin)
+	err = searchAndDetach(name, cgroupRoot, nil)
+	// Treat detaching unsupported attach types as successful.
+	if errors.Is(err, link.ErrNotSupported) {
+		return nil
+	}
+	return err
+}
+
+// tryPin pins the link to maintain it after agent exit. l is closed after
+// a successful pin operation.
+//
+// Returns ErrNotSupported if l was attached using PROG_ATTACH.
+func tryPin(l link.Link, name, pinPath string) error {
+	err := l.Pin(filepath.Join(pinPath, name))
+	if errors.Is(err, link.ErrNotSupported) {
+		log.Debugf("program %s was attached using PROG_ATTACH", name)
+	}
+	if err != nil {
+		return fmt.Errorf("pin link: %w", err)
+	}
+
+	log.Debugf("program %s attached using bpf_link", name)
+	return l.Close()
+}
+
+// searchAndDetach detaches all programs attached to cgroupRoot with the
+// corresponding attach type for a program named 'name'. If skip is given,
+// that program will be skipped during the operation.
+func searchAndDetach(name, cgroupRoot string, skip *ebpf.Program) error {
+	fd, err := os.Open(cgroupRoot)
+	if err != nil {
+		return fmt.Errorf("open cgroup %s: %w", cgroupRoot, err)
+	}
+	defer fd.Close()
+
+	attach := attachTypes[name]
+	ids, err := link.QueryPrograms(link.QueryOptions{
+		Path:   cgroupRoot,
+		Attach: attach,
+	})
+	// We know the cgroup root exists, so EINVAL will likely mean querying
+	// the given attach type is not supported.
+	if errors.Is(err, unix.EINVAL) {
```

Comment:
> this function and this could end up causing us to mistake a true EINVAL for ENOTSUPP

We've worked on feature probes a ton, and this is just how the sausage is made. You always need to proceed with known-good inputs (like the cgroup root here; we're 100% sure it exists as we have an open fd to it) and work by elimination and deduction. We know `Path` is valid, so if we receive `EINVAL`, that must mean `Attach` is not. (going by the kernel source, that's what EINVAL means given these particular inputs)

This kernel API is in maintenance mode as it's been replaced by bpf_link, so barring any significant changes upstream, this shouldn't break. And if it does, other probes will break too, so upstream is generally careful with these changes.

> if the common EINVAL failure case here today is that the feature is not supported, then it's probably more user-friendly to wrap that and suggest to the user that this is the reason for the failure.

That's exactly what we're doing here, right? `err` (the original error) is included as a string for humans to read, while `ErrNotSupported` is a sentinel other code can act on.

Reply:
(yes on the last point, I was trying to acknowledge that the best I could come up with is exactly what you've done, so this discussion doesn't require any action. :+1: )